service/history/shard/controller_impl.go

```diff
@@ -250,47 +210,46 @@ func (c *ControllerImpl) getOrCreateHistoryShardItem(shardID int32) (*historySha
 		return nil, err
 	}
 
-	if info.Identity() == c.GetHostInfo().Identity() {
-		shardItem, err := newHistoryShardsItem(
-			c.Resource,
-			shardID,
-			c.engineFactory,
-			c.config,
-		)
-		if err != nil {
-			return nil, err
-		}
-		c.historyShards[shardID] = shardItem
-		c.metricsScope.IncCounter(metrics.ShardItemCreatedCounter)
+	if info.Identity() != c.GetHostInfo().Identity() {
```

Comment:
can you move this line and above Lookup function call before `c.Lock()`?

Reply:
GetHostInfo may invoke network layer, meaning it can be relatively expensive.

Reply:
As far as I can tell, it only does that at startup. `GetHostInfo().Identity()` does not do any network calls (or function calls at all). Anyway, if that were true, then we should cache the value in the controller and context, since we use it a bunch.